sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java

```diff
@@ -88,108 +100,220 @@ public static void main(String[] args) throws IOException, InterruptedException
-                client.listIncomingRelationships(twinId)
-                    .doOnComplete(deleteRelationshipsSemaphore::release)
-                    .doOnError(throwable -> {
-                        if (throwable instanceof ErrorResponseException && ((ErrorResponseException) throwable).getResponse().getStatusCode() == HttpStatus.SC_NOT_FOUND) {
-                            deleteRelationshipsSemaphore.release();
-                        } else {
-                            System.err.println("List incoming relationships error: " + throwable);
-                        }
-                    })
-                    .subscribe(
-                        incomingRelationship -> client.deleteRelationship(incomingRelationship.getSourceId(), incomingRelationship.getRelationshipId())
-                            .subscribe(
-                                aVoid -> System.out.println("Found and deleted incoming relationship: " + incomingRelationship.getRelationshipId()),
-                                throwable -> System.err.println("Delete incoming relationship error: " + throwable)
-                            ));
+        // Once the async API terminates (either successfully, or with an error), the latch count is decremented, or the semaphore is released.
+        for (Map.Entry<String, String> twin : twins.entrySet()) {
+            String twinId = twin.getKey();
+
+            List<BasicRelationship> relationshipList = new ArrayList<>();
+            Semaphore listRelationshipSemaphore = new Semaphore(0);
+            Semaphore deleteRelationshipsSemaphore = new Semaphore(0);
+            CountDownLatch deleteTwinsLatch = new CountDownLatch(1);
+
+            // Call APIs to retrieve all relationships.
+            client.listRelationships(twinId, BasicRelationship.class)
+                .doOnNext(relationshipList::add)
+                .doOnError(IgnoreNotFoundError)
+                .doOnTerminate(listRelationshipSemaphore::release)
+                .subscribe();
+
+            // Call APIs to retrieve all incoming relationships.
+            client.listIncomingRelationships(twinId)
+                .doOnNext(e -> relationshipList.add(mapper.convertValue(e, BasicRelationship.class)))
+                .doOnError(IgnoreNotFoundError)
+                .doOnTerminate(listRelationshipSemaphore::release)
+                .subscribe();
+
+            // Call APIs to delete all relationships.
+            if (listRelationshipSemaphore.tryAcquire(2, MaxWaitTimeAsyncOperationsInSeconds, TimeUnit.SECONDS)) {
+                relationshipList
+                    .forEach(relationship -> client.deleteRelationship(relationship.getSourceId(), relationship.getId())
+                        .doOnSuccess(aVoid -> {
+                            if (twinId.equals(relationship.getSourceId())) {
+                                System.out.println("Found and deleted relationship: " + relationship.getId());
```

Comment:
Does Java support string interpolation?

Reply:
We can use `String.format()` to specify a template, but nothing as handy as `$` in C#.